### PR TITLE
[release/2.4] Cherry-picks from release/2.3

### PR DESCRIPTION
**Pull request by dnikolaev-amd**
Cherry-pick details here: https://amdcloud-my.sharepoint.com/:x:/r/personal/pmagundu_amd_com/Documents/pytorch/ROCm_fork_pytorch_branch_diff.xlsx?d=w2c1b2bbc6f314cc38e97be74d273821d&csf=1&web=1&e=QRboKB&nav=MTVfe0IxMTU0RjkxLTk4NkEtNEVGMS1CNEU3LUNBRTc5N0Q4MkFGM30

### CODE REVIEW

File: test/functorch/test_ops.py [hunk 1]
```diff
@@ -1021,23 +1021,21 @@ def fn(inp, *args, **kwargs):
...
-        ),
-    )
-    @skipOps(
-        "TestOperators",
-        "test_vmapvjpvjp",
-        {
-            xfail("as_strided", "partial_views"),
-        },
-    )
+    @opsToleranceOverride('TestOperators', 'test_vmapvjpvjp', (
```
Comment (pruthvistony) [inline test/functorch/test_ops.py:1024]: This indentation change is NOT required, revert this change.

File: test/inductor/test_select_algorithm.py [hunk 2]
```diff
@@ -109,7 +109,6 @@ def foo(a, b):
         )
         self.assertEqual(counters["inductor"]["select_algorithm_autotune"], 1)
 
-    # FIXME: Investigate why _int_mm_out_cuda is not compiled on ROCm
```
Comment (pruthvistony) [inline test/inductor/test_select_algorithm.py:112]: Why this is removed.

File: test/test_matmul_cuda.py [hunk 3]
```diff
@@ -374,21 +374,24 @@ def _test_tautological_mm(self, device: str = "cuda",
     @unittest.skipIf(not scaled_mm_supported_device(), f8_msg)
     def test_float8_basics(self, device) -> None:
         self._test_tautological_mm(device, e4m3_type, e4m3_type, size=16)
-        # hipblaslt does not yet support mixed e4m3_type input
         if torch.version.hip is None:
-            self._test_tautological_mm(device, e4m3_type, e5m2_type, size=32)
-            self._test_tautological_mm(device, e5m2_type, e4m3_type, size=48)
-        # According to https://docs.nvidia.com/cuda/cublas/#id99 8F_E5M2 MM is unsupported
-        with self.assertRaises(RuntimeError):
+            # According to https://docs.nvidia.com/cuda/cublas/#id99 8F_E5M2 MM is unsupported
```
Comment (pruthvistony) [inline test/test_matmul_cuda.py:378]: why the CUDA case is change? how did this change come in? due to git merge?

File: test/test_matmul_cuda.py [hunk 4]
```diff
@@ -374,21 +374,24 @@ def _test_tautological_mm(self, device: str = "cuda",
...
             self._test_tautological_mm(device, e5m2_type, e5m2_type)
 
+
+        self._test_tautological_mm(device, e4m3_type, e5m2_type, size=32)
+        self._test_tautological_mm(device, e5m2_type, e4m3_type, size=48)
+
         self._test_tautological_mm(device, size=64, out_dtype=torch.float16)
         self._test_tautological_mm(device, size=96, out_dtype=torch.float32)
-        # hipblaslt does not yet support bfloat16 output
+        self._test_tautological_mm(device, size=80, out_dtype=torch.bfloat16)
```
Comment (pruthvistony) [inline test/test_matmul_cuda.py:390]: same comment as above.

File: test/test_matmul_cuda.py [hunk 5]
```diff
@@ -374,21 +374,24 @@ def _test_tautological_mm(self, device: str = "cuda",
...
+
         if torch.version.hip is None:
-            self._test_tautological_mm(device, size=80, out_dtype=torch.bfloat16)
-        with self.assertRaises(RuntimeError):
-            self._test_tautological_mm(device, out_dtype=e5m2_type)
+            with self.assertRaises(RuntimeError):
```
Comment (pruthvistony) [inline test/test_matmul_cuda.py:393]: same comment as above

**Review by pruthvistony — CHANGES_REQUESTED** [review]

File: test/functorch/test_ops.py (hunk 1 above)
Comment (dnikolaev-amd) [reply test/functorch/test_ops.py:1024]: we have some PRs for different branches to extend tolerance in  `test/functorch/test_ops.py:test_vmapvjpvjp_linalg_householder_product_cuda_float32` 
https://github.com/ROCm/frameworks-internal/issues/7750

File: test/test_matmul_cuda.py (hunk 3 above)
Comment (dnikolaev-amd) [reply test/test_matmul_cuda.py:378]: here is some rearrange for tests. 
some of tests are passed on rocm now: 
source: https://github.com/ROCm/pytorch/commit/9f765703749d5c7744433241057f0df9601b8798

Comment (dnikolaev-amd) [reply test/test_matmul_cuda.py:378]: cherry-picked from https://github.com/ROCm/pytorch/commit/9521966dfed08c968507888a26f0c502a576d9a9

File: test/inductor/test_select_algorithm.py (hunk 2 above)
Comment (dnikolaev-amd) [reply test/inductor/test_select_algorithm.py:112]: This changes was made when resolving the conflict with commit https://github.com/ROCm/pytorch/commit/9521966dfed08c968507888a26f0c502a576d9a9

File: test/functorch/test_ops.py (hunk 1 above)
Comment (dnikolaev-amd) [reply test/functorch/test_ops.py:1024]: Comes from commit https://github.com/ROCm/pytorch/commit/9521966dfed08c968507888a26f0c502a576d9a9 to adjust tolerance  for `test/functorch/test_ops.py:test_vmapvjpvjp_linalg_householder_product_cuda_float32`

**Review by pruthvistony — APPROVED** [review]
